src/doc/design_overview/overview.rst

```diff
@@ -0,0 +1,961 @@
+     \|/       \|/      \|/      \|/
+     Hub<----> Hub<---->Hub<---> Hub  <-----------> RethinkDB <--> RethinkDB  <--> RethinkDB ...
+               /|\      /|\      /|\
+                |        |        |
+       ---------|        |        | (tcp)
+       |                 |        |
+       |                 |        |
+      \|/               \|/      \|/
+     Compute<-------->Compute<-->Compute <--- rsync replication  to Storage Server, which has BTRFS snapshots
+
+(with the exception that William is in the process of replacing RethinkDB
+with PostgreSQL).
+
+It may be helpful to explain some of the entities in this diagram a bit
+more.
+
+Client
+^^^^^^
+
+This is the SMC client interface, built primarily with React and bundled
+together webpack as described previously.  When a user goes to the root of SMC
+in their web browser, the HAproxy configuration serves it the ``index.html``
+from its default backend, which happens to be a simple nginx server dedicated
+to static files.  It also gets images, and the client JavaScript from the
+static server.  Once the JavaScript takes over everything else happens in the
+browser including setting up the appropriate view for the client (whether or
+not they're logged in, etc.) and communicating with the hub using
+Primus/Engine.IO (through the HAproxy--more on that next).  The majority of the
+client is implemented in the code in ``smc-webapp`` and ``webapp-lib``, with
+some bits from ``smc-util``.
+
+HAproxy
+^^^^^^^
+
+HAproxy serves as the front line to all connections from clients to SMC.  It
+routes all connections to different backends depending, primarily, on the
+URL (and port).  The main frontend is of course HTTPS over port 443.  By
+default requests are sent to the static file server (nginx) as mentioned
+above.  Most other requests are sent to the 'hub' backend, which may be
+running any number of the hub servers, one of which is selected using the
+currently configured load balancing scheme (it also uses a session cookie to
+keep individual clients connected to the same hub instance).
+
+Hub
+^^^
+
+The "Hub" is the primary server backend for SMC, built on Node.js as described
+previously.  It consists of an HTTP server with Primus + Engine.IO attached to
+handle real-time bidirectional client/hub communication.  Most communication
+between the Client and the backend happens through the Hub, whose HTTP server
+uses Express to route requests to different services (account management,
+project management, payment, etc.).  Each Hub also sets up a
+``ComputeServerClient`` which gives it access to all the running compute
+servers (discussed next).  The names and URLs of all the available compute
+servers live in a system table in the database.
+
+It also uses `node-http-proxy`_ to create an HTTP proxy server associated
+with each Hub (on port number one higher than the Hub's HTTP port).  If we
+understand correctly, the proxy handles all requests that are to be
```

Comment:
Right, I'll mention that.  I know how that works.  At some point it would be good to generalize how the Jupyter server is handled too.